src/collective/communicator_factory.cu

```diff
@@ -0,0 +1,77 @@
+/*!
+ * Copyright 2022 XGBoost contributors
+ */
+#include "communicator_factory.h"
+#include "device_communicator_adapter.cuh"
+#include "rabit_communicator.h"
+#ifdef XGBOOST_USE_NCCL
+#include "nccl_device_communicator.cuh"
+#endif
+#if defined(XGBOOST_USE_FEDERATED)
+#include "../../plugin/federated/federated_communicator.h"
+#endif
+
+namespace xgboost {
+namespace collective {
+
+thread_local std::unique_ptr<CommunicatorFactory> CommunicatorFactory::instance_{};
+
+CommunicatorFactory::CommunicatorFactory(CommunicatorType type, Communicator* communicator)
```

Comment:
Maybe we can use dmlc registry to help define the factory?

Reply:
Yeah I think it's possible, but we need to figure out how to configure the parameters first. Will come back to this.

Reply:
Ended up not using the registry since we need some logic to pick which communicator to use, but got rid of the factories.